### ISSUE TITLE
MoltenVK 1.2.0 performance regression on AMD gpu

### ISSUE THREAD

**Issue by Qotonana** [KhronosGroup/MoltenVK#1746]
Hi all

First, congrats on this project to all involved.
Doing some testing with CrossOver 22 and new MoltenVK (unsupported off course :)) and noticed that the new MoltenVK-1.2.0 has terrible performance regression on Intel Mac with AMD Radeon card.

Uses libMoltenVK dylibs from @Gcenx releases (just repackaged) from
https://github.com/The-Wineskin-Project/MoltenVK/releases/
Tested on Monterey 12.6. Cleared the DXVK state cache before every test. 

Also replaced DXVK with macOS-async-1.10.3 from
https://github.com/Gcenx/DXVK-macOS/releases

**Sniper Contracts MoltenVK macos-1.1.11**
https://github.com/The-Wineskin-Project/MoltenVK/releases/download/v1.1.11/macos-1.1.11.tar.xz
![Sniper Contracts macos-1 1 11](https://user-images.githubusercontent.com/102525166/196627138-f892d7e3-c3d9-4ea4-8634-efdf58b07aee.png)

**Sniper Contracts MoltenVK macOS-1.2.0**
https://github.com/The-Wineskin-Project/MoltenVK/releases/download/v1.2.0/macos-1.2.0.tar.xz
![Sniper Contracts macos-1 2 0](https://user-images.githubusercontent.com/102525166/196627227-5e253f3f-8a02-4dec-90e1-8da9b9826c06.png)

**Automobilista 2 DemoAVX MoltenVK macos-1.1.11**
![Automobilista 2 DemoAVX 2560x1440 DXVK113async macos-1 1 11](https://user-images.githubusercontent.com/102525166/196627306-fee5513c-d723-4576-bfa7-51d32f9e8391.jpg)

**Automobilista 2 DemoAVX MoltenVK macos-1.2.0**
![Automobilista 2 DemoAVX 2560x1440 DXVK113async macos-1 2 0-beta](https://user-images.githubusercontent.com/102525166/196627377-a3258d14-5cf9-440e-88c4-38e750f6e86a.jpg)

Tried also to add various env variables and getting the same results
"MVK_ALLOW_METAL_FENCES" = "1"
"MVK_ALLOW_METAL_EVENTS" = "1"
"MVK_CONFIG_VK_SEMAPHORE_SUPPORT_STYLE" = "MVK_CONFIG_VK_SEMAPHORE_SUPPORT_STYLE_METAL_EVENTS"

And sorry if this needs to go to Discussions forum 
Thanks

**Comment by phush0** [comment]
it should be `"MVK_CONFIG_VK_SEMAPHORE_SUPPORT_STYLE" = "2"'

**Comment by Gcenx** [comment]
> it should be `"MVK_CONFIG_VK_SEMAPHORE_SUPPORT_STYLE" = "2"`

This was attempted and secondly using the full name this would also be acceptable see
https://github.com/KhronosGroup/MoltenVK/blob/260bad42e1fbcc10afb2ddf870487df6dd895e01/MoltenVK/MoltenVK/Utility/MVKEnvironment.h#L238

<br>

Before anyone asks the linked packages are stock builds taken from this repository’s GitHub action runs just repackaged into tar.xz archives.

**Comment by billhollings** [comment]
Was there any further research done on this? Did it turn out to be the semaphore style you were using?

Has any profiling been done to see where the bottlenecks are?

Is there anyway we can demonstrate this here, to debug?

**Comment by Qotonana** [comment]
> Was there any further research done on this? Did it turn out to be the semaphore style you were using?
> 
> Has any profiling been done to see where the bottlenecks are?
> 
> Is there anyway we can demonstrate this here, to debug?

Just retested all again


Steps to reproduce bug

_CrossOver 22.0.1_

Clean Steam bottle without any EnvironmentVariables, except "DXVK_HUD" = "full" for performance comparisons, added to cxbottle.conf ( _/Users/.../Library/Application Support/CrossOver/Bottles/Steam_ )

DXVK, CSMT and Esync all on 
iMac Intel with 570x AMD gpu

Replaced the default CrossOver build of mvk libMoltenVK.dylib from
_/Applications/CrossOver.app/Contents/SharedSupport/CrossOver/lib64_

With the newer one from
_https://github.com/The-Wineskin-Project/MoltenVK/releases/_ ( Release for Vulkan SDK 1.3.231 (macos-1.2.0.) )

Also tried different libMoltenVK.dylib from _https://github.com/KhronosGroup/MoltenVK/actions_ and any newer one than MoltenVK Release for Vulkan SDK 1.3.224 (the last one that don't tank performance) resulted in almost half fps in every games tested.

Notes
- This method of "updating MoltenVK" in CrossOver is of course unsupported by Codeweavers, but used by many users as MoltenVK updates constantly improve performance and games compatibility (at least until SDK 1.3.231 on Mac with AMD gpu card)
- I can assume similiar regression in other apps that use MoltenVK 1.2 but didn't tested further 

Thank You

**Comment by billhollings** [comment]
> Steps to reproduce bug

Downloaded CrossOver, configured as you indicate. I downloaded the free _Automobilista 2 Demo VW TSI Cup_.

I can get it running on Crossover/Steam on my Intel/AMD Mac. According to the DXVK HUD, it runs at about 35ms per frame.

However, I cannot find a way to substitute in a more recent version of `libMoltenVK.dylib`, so I can evaluate the difference, and debug. In attempting to do so, I consistently get the following error when launching the game. I must be missing something. Help would be appreciated.

![Screen Shot 2022-12-02 at 8 39 58 AM](https://user-images.githubusercontent.com/356402/205320375-2b0777c1-e541-4e95-b1c9-56932f0b31c8.jpg)

**Comment by Gcenx** [comment]
To use completely stock versions of libMoltenVK you’ll need to swap out the DXVK dlls, firstly close CrossOver completely

The DXVK dlls are found
```
CrossOver.app/Contents/SharedSupport/CrossOver/lib64/wine/dxvk
```
Delete everything inside that directory and replace them with the dlls from https://github.com/Gcenx/DXVK-macOS/releases/download/v1.10.3/dxvk-macOS-async-1.10.3_1.tar.gz

Now open CrossOver and turn off DXVK and enable it again so it copies in the newer dlls.

**Comment by billhollings** [comment]
Slowly getting there.

I've got it running with latest MoltenVK, and can confirm a reduced FPS.

I'm trying to attach _Instruments_ to the process, to investigate bottlenecks. There doesn't seem to be any process with the game name. The Mac Dock shows the AMS game as `wine64-preloader`, but there are MANY `wine64-preloader` processes active. Attaching Instruments to the busiest of these does not yield any recognizable MoltenVK calls in the call stacks within _Instruments_.

Is there a mechanism for reliably determining the process that is running the game graphics?

**Comment by Gcenx** [comment]
None that I’m aware of unfortunately maybe @cdavis5e knows of a reliable way.

**Comment by cdavis5e** [comment]
There's a system-private interface (SPI) in Launch Services that can set the name that shows up in the Dock and in Activity Monitor, but I don't know the details. I only know it exists because WebKit uses it for its subprocesses. The details of how to call it are hidden behind a binary that is checked into the WebKit repo and linked to WebKit at link time, so you can't go look at the WebKit source and go use this. I'm kind of disappointed, because I really wanted to use that in Wine so I don't have to wade through 50 different `wine` processes to find the one that I'm interested in.

Normally what I do is, in Activity Monitor, I open the process information panel for each of those `wine{64,}-preloader` processes and look for the process that has `mmap(2)`'d the game's main binary--it'll show up on the "Open Files and Ports" tab. Or just use `lsof(8)`, which is what Activity Monitor does to populate that tab. Or use `ps(1)`--Wine sets up its command line so that `wine` doesn't show up in the command line; it'll look like you ran the `*.exe` file directly.

**Comment by billhollings** [comment]
Fixed in PR #1793. Please retest using latest MoltenVK code, and close this issue if the problem is fixed.

**Comment by Qotonana** [comment]
I retested and can confirm that issue is fixed in PR https://github.com/KhronosGroup/MoltenVK/pull/1793.

Thank You
